src/index/reorg.rs

```diff
@@ -0,0 +1,98 @@
+use {super::*, updater::BlockData};
+
+#[derive(Debug, PartialEq)]
+pub(crate) enum ReorgError {
+  Recoverable((u64, u64)),
+  Unrecoverable,
+}
+
+impl fmt::Display for ReorgError {
+  fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+    match self {
+      ReorgError::Recoverable((height, depth)) => {
+        write!(f, "{depth} block deep reorg detected at height {height}")
+      }
+      ReorgError::Unrecoverable => write!(f, "unrecoverable reorg detected"),
+    }
+  }
+}
+
+impl std::error::Error for ReorgError {}
+
+const MAX_SAVEPOINTS: usize = 2;
+const SAVEPOINT_INTERVAL: u64 = 5;
+
+pub(crate) struct Reorg {}
+
+impl Reorg {
+  pub(crate) fn detect_reorg(block: &BlockData, height: u64, index: &Index) -> Result {
+    let bitcoind_prev_blockhash = block.header.prev_blockhash;
+
+    match index.block_hash(height.checked_sub(1))? {
+      Some(index_prev_blockhash) if index_prev_blockhash == bitcoind_prev_blockhash => Ok(()),
+      Some(index_prev_blockhash) if index_prev_blockhash != bitcoind_prev_blockhash => {
+        let max_recoverable_reorg_depth =
+          height % SAVEPOINT_INTERVAL + (MAX_SAVEPOINTS as u64 * SAVEPOINT_INTERVAL);
+
+        for depth in 1..max_recoverable_reorg_depth {
+          let index_block_hash = index.block_hash(height.checked_sub(depth))?;
+          let bitcoind_block_hash = index
+            .client
+            .get_block_hash(height.saturating_sub(depth))
+            .into_option()?;
+
+          if index_block_hash == bitcoind_block_hash {
+            return Err(anyhow!(ReorgError::Recoverable((depth, height))));
+          }
+        }
+
+        Err(anyhow!(ReorgError::Unrecoverable))
+      }
+      _ => Ok(()),
+    }
+  }
+
+  pub(crate) fn handle_reorg(index: &Index, height: u64, depth: u64) -> Result {
+    log::info!("rolling back database after reorg of depth {depth} at height {height}");
+
+    let mut wtx = index.begin_write()?;
+
+    let oldest_savepoint =
```

Comment:
For this, an alternative to consider is to rollback to the latest savepoint and delete it. That would only roll you back 1-4 blocks instead of 10-14. If it's not far enough, then when the updater runs again it will hit a reorg again and rollback to the older savepoint automatically.